pulsar-client/src/main/java/com/yahoo/pulsar/client/impl/ConsumerImpl.java

```diff
@@ -526,6 +571,28 @@ void connectionOpened(final ClientCnx cnx) {
     }
 
     /**
+     * Clear the internal receiver queue and returns the message id of what was the 1st message in the queue that was
+     * not seen by the application
+     */
+    private MessageIdImpl clearReceiverQueue() {
+        List<Message> currentMessageQueue = new ArrayList<>(incomingMessages.size());
+        incomingMessages.drainTo(currentMessageQueue);
+        if (!currentMessageQueue.isEmpty()) {
+            MessageIdImpl nextMessageInQueue = (MessageIdImpl) currentMessageQueue.get(0).getMessageId();
+            MessageIdImpl previousMessage = new MessageIdImpl(nextMessageInQueue.getLedgerId(),
+                    nextMessageInQueue.getEntryId() - 1, nextMessageInQueue.getPartitionIndex());
+            return previousMessage;
+        } else if (lastDequeuedMessage != null) {
+            // If the queue was empty we need to restart from the message just after the last one that has been dequeued
+            // in the past
+            return lastDequeuedMessage;
```

Comment:
> So, is there any issue if reader starts with lastDequeuedMessage=2 or just takes incomingMessages.poll(0, TimeUnit.MILLISECONDS)-1 and then incomingMessages.clear()

The goal here is to have exact precision on which message to ask the broker after the reconnection. When we drain the queue, we can determine with (with no race conditions) which is the 1st message that was in the queue. 

If we just rely on the lastDequeuedMessage, the dequeing and the update of the lastDequeuedMessage are not synchronized and thus that could lead to some re-deliveries.